packages/vite/src/node/plugins/resolve.ts

```diff
@@ -583,43 +592,24 @@ function tryCleanFsResolve(
         // try resolve .js, .mjs, .cjs or .jsx import to typescript file
         const fileExt = path.extname(file)
         const fileName = file.slice(0, -fileExt.length)
-        if (
-          (res = tryResolveRealFile(
-            fileName + fileExt.replace('js', 'ts'),
-            preserveSymlinks,
-          ))
-        )
-          return res
+        if ((res = tryResolveFile(fileName + fileExt.replace('js', 'ts'))))
+          return getRealPath(res, options.preserveSymlinks)
```

Comment:
Is there a reason we're changing the pattern for when the real path is resolved? I kinda prefer the old way as the function would return the finalized resolved path directly, but maybe I'm missing an optimization made here.

Reply:
See https://github.com/vitejs/vite/pull/15211/commits/0faf06c0cb4f9abb5ada86f2b1eee59cf33d900e, doing it in this way avoids creating many promises in this function.

But I don't think we will be doing it this way if we merge https://github.com/vitejs/vite/pull/15279. I'm unsure whether generating all the promises is justified. readdirSync takes almost no time compared to other things, and we could expand the tree in parallel (but keeping get/set sync). There are some fsp.readFile in this PR though. I think we need to reevaluate and maybe break this one in pieces after #15279

Reply:
Why won't we try to avoid creating 10+ promises for each `tryCleanFsResolve` 🤔
Doing an await for them in a for loop isn't free either. We have merged PRs to move from a `.startsWith('/')` to `[0] === '/'`. I would imagine this has a way bigger effect, more thinking on the garbage collector, etc. But I think we will revert this change if we want to merge this one.

About populating the fs tree cache async, ya, we need to review what works better.

Reply:
Hmm, I can't argue that no await is definitely faster that some awaits 😄 I don't mind as much now looking more at it, but initially I didn't think it was worth the perf. `startsWith` was different that it was an easy change, otherwise I would keep `startsWith` for readability.